### PR TITLE
Fix CmaEs system attribution key

### PR DESCRIPTION
**Pull request by gen740**
## Motivation

Related to #4016 and #4142.

As noted in #4142, CmaEs's trial attribution is not properly working. Keys are different in Setting and Getting. This PR will fix this issue.

## Description of the changes

- Add a new private property method `_attr_keys` to `CmaEsSampler`
- Using `NamedTuple` to confirm the `_attr_keys`'s key
- Let `CmaEsSampler`'s `_concat_optimizer_attrs` and `_split_optimizer_str` member function, since it require `use_separable_cma` and `with_margin` information.
- Rewrite cmaes_test. (This is WIP)

Could anyone explain what [backward compatibility](https://github.com/gen740/optuna/blob/master/optuna/samplers/_cmaes.py#L452-L456) means?

### CODE REVIEW

**Comment by c-bata** [comment]
@eukaryo Could you review this PR after passed CI?

**Comment by eukaryo** [comment]
Sorry for the late reply. I am OK to be assigned as a reviewer. However, I think it would be better to have another reviewer with more experience.

File: optuna/samplers/_cmaes.py
Comment (c-bata) [inline optuna/samplers/_cmaes.py:46]: The following code was prepared for users who used `CmaEsSampler` in prior to Optuna v2.5, but it seems old enough, so what do you think about removing it in this PR?

https://github.com/optuna/optuna/blob/8bef4de77284965a805560030f169eaa7efdcaf9/optuna/samplers/_cmaes.py#L490-L496

**Review by c-bata — COMMENTED** [review]
Thank you for your pull request. I left two questions.

File: optuna/samplers/_cmaes.py
Comment (gen740) [reply optuna/samplers/_cmaes.py:455]: Since `_restore_optimizer` returns 0 for n restarts. I doubt this attr_key is doing anything, or even worse this function seems not to restore the optimizer correctly.
https://github.com/optuna/optuna/blob/8bef4de77284965a805560030f169eaa7efdcaf9/optuna/samplers/_cmaes.py#L502

Comment (gen740) [reply optuna/samplers/_cmaes.py:455]: Is it OK for `attr_keys` not to contains `restart_n` information?

**Review by c-bata — CHANGES_REQUESTED** [review]
Sorry, please ignore this comment since the following test case is incorrect.

<details>

There seems not to be enough test case to check this change. For example, I add following scenario, which is failed.

```python
@pytest.mark.parametrize("sampler_opts", [
    {"use_separable_cma": False, "with_margin": False},
    {"use_separable_cma": True, "with_margin": False},
    {"use_separable_cma": False, "with_margin": True},
])
def test_restore_optimizer_from_substrings(sampler_opts: Dict[str, Any]) -> None:
    sampler = optuna.samplers.CmaEsSampler(popsize=8, **sampler_opts)
    optimizer, n_restarts = sampler._restore_optimizer([])
    assert optimizer is None
    assert n_restarts == 0

    def objective(trial: optuna.Trial) -> float:
        x1 = trial.suggest_float("x1", -10, 10)
        x2 = trial.suggest_float("x2", -10, 10)
        return x1 ** 2 + x2 ** 2

    study = optuna.create_study(sampler=sampler)
    study.optimize(objective, 10)
    completed_trials = study.get_trials(states=[TrialState.COMPLETE])

    optimizer, n_restarts = sampler._restore_optimizer(completed_trials)

    assert optimizer is not None
    assert n_restarts == 1
    if sampler._with_margin:
        assert isinstance(optimizer, CMAwM)
    elif sampler._use_separable_cma:
        assert isinstance(optimizer, SepCMA)
    else:
        assert isinstance(optimizer, CMA)
```

Let me try to add more test scenarios :bow:

</details>

File: optuna/samplers/_cmaes.py
Comment (c-bata) [reply optuna/samplers/_cmaes.py:46]: OK. Let me remove it in the following pull request.

File: optuna/samplers/_cmaes.py
Comment (c-bata) [reply optuna/samplers/_cmaes.py:455]: > Since `_restore_optimizer` returns 0 for n restarts. I doubt this attr_key is doing anything, or even worse this function seems not to restore the optimizer correctly.

`return None, 0` means that the optimizer must be initialized. So the logic seems to be correct.

Comment (c-bata) [reply optuna/samplers/_cmaes.py:455]: I added some test cases to check this logic and it works expected even after restarted. Let me resolve this comment.
https://github.com/gen740/optuna/compare/cmaes-sys-attr-key...c-bata:optuna:cmaes-sys-attr-key?expand=1

**Review by c-bata — APPROVED** [review]
Thank you for your pull request.
Although there seems not to be enough test cases, I added some test scenarios and confirmed that the current logic passed all scenarios.
https://github.com/gen740/optuna/compare/cmaes-sys-attr-key...c-bata:optuna:cmaes-sys-attr-key?expand=1

Let me add the test scenarios at #4233.
LGTM.

**Comment by c-bata** [comment]
Let me merge this PR after you approved :pray:

**Review by eukaryo — APPROVED** [review]
Sorry for the late reply. LGTM

**Pull request merged**
